### PR TITLE
Metainfo user names

### LINKED ISSUE #12787: [Improvement]: Show user's name on metadata information panel
### Improvement description

When viewing asset metadata Pimcore will show the user ID of the owner and user who last modified the asset. Administrators have the ability to click on the user ID to view the user's details but this is not available to all users.

When there are multiple users of the DAM, it is nice to have the full name of the user shown in the panel, not just the ID.

### PR DESCRIPTION
**Pull request by Andreas-Koepke**
Add new endpoint for getting full name of user, supports multiple users.
Add full name of user to metadata panel.
  
Resolves #12787

### CODE REVIEW

**Comment by Andreas-Koepke** [comment]
This ended up being a much more significant change than expected. 

I have updated all of the associated controllers to now include the user's name. 

All of the JS files that use pimcore.element.metainfo have also been updated.

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php [hunk 1]
```diff
@@ -479,6 +480,24 @@ static function (Task $task) {
             $objectData['general']['versionDate'] = $objectFromDatabase->getModificationDate();
             $objectData['general']['versionCount'] = $objectFromDatabase->getVersionCount();
 
+            $userOwner = User::getById($objectData['general']['o_userOwner']);
+            if (empty($userOwner)) {
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php]: I think we should make this a bit more generic, what about adding a `getFullName()` method in the `User` object and maybe a trait that handles all the displaying related stuff?

**Review by brusch — CHANGES_REQUESTED** [review]

**Comment by aryaantony92** [comment]
@Andreas-Koepke Thank you for the PR. And just a gentle reminder about the generic approach. Please refer to the comment above.

**Comment by Andreas-Koepke** [comment]
Sounds good to me. I can make these changes later this week.

**Comment by aryaantony92** [comment]
> Sounds good to me. I can make these changes later this week.

Thank you 👍

**Comment by Andreas-Koepke** [comment]
I have added in the new method for the User object and updated my code.

As for the trait for the display related code, I am not sure exactly what part of my changes you are referring to.

File: bundles/AdminBundle/Controller/Admin/Asset/AssetController.php [hunk 2]
```diff
@@ -217,6 +218,24 @@ static function (Task $task) {
...
+
+        $userModification = ($asset->getUserOwner() == $asset->getUserModification()) ? $userOwner : User::getById($asset->getUserModification());
+        if (empty($userModification)) {
+            $data['userModificationUsername'] = '';
+            $data['userModificationFullname'] = $this->trans('user_unknown');
+        } else {
+            $data['userModificationUsername'] = $userModification->getName();
+            $data['userModificationFullname'] = $userModification->getFullName();
+        }
+
```
Comment (aryaantony92) [inline bundles/AdminBundle/Controller/Admin/Asset/AssetController.php]: We have similar code (for username and usermodification) in DataObjectController, DocumentController etc. 
If we add this code in a trait, then different controllers can access it from there. Thanks!

Comment (Andreas-Koepke) [reply bundles/AdminBundle/Controller/Admin/Asset/AssetController.php]: The new trait has been added, please review the updated code.

I have also made a fix to the metainfo panel so it handles the system account properly.

Comment (aryaantony92) [reply bundles/AdminBundle/Controller/Admin/Asset/AssetController.php]: Thank you! . Could you please resolve the conflicts also ?

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * @internal
+ */
+trait UserNameTrait
+{
+    /**
+     * @param int $user_id The User ID.
+     */
+    protected function getUserName($user_id) {
```
Comment (blankse) [inline bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]:
```suggestion
    /**
     * @param int $user_id The User ID.
     *
     * @return array{username: string, fullname: string}
     */
    protected function getUserName(int $user_id): array
    {
```

We should use type hints for new methods

File: models/User.php [hunk 4]
```diff
@@ -216,6 +216,14 @@ public function setLastname($lastname)
         return $this;
     }
 
+    /**
+     * @return string|null
+     */
+    public function getFullName()
```
Comment (blankse) [inline models/User.php]:
```suggestion
    public function getFullName(): string
```
We should use type hints for new methods

File: bundles/AdminBundle/src/Controller/Admin/Document/DocumentControllerBase.php [hunk 5]
```diff
@@ -465,4 +468,18 @@ protected function saveDocument(Model\Document $document, Request $request, bool
 
         return [$task, $document, $version];
     }
+
+    /**
+     * @param Model\Document $document
+     * @param array $data
+     */
```
Comment (blankse) [inline bundles/AdminBundle/src/Controller/Admin/Document/DocumentControllerBase.php]:
```suggestion
```
We should avoid redundant php docs

**Comment by Andreas-Koepke** [comment]
Merge conflicts have been resolved and some final modifications made.

**Comment by aryaantony92** [comment]
> Merge conflicts have been resolved and some final modifications made.

Thank you.

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Bundle\AdminBundle\Controller\Traits;
+
+use Pimcore\Model\User;
+
+/**
+ * @internal
+ */
```
Comment (NiklasBr) [inline bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:22]: Why `internal`? Seems like a useful Trait which could be used elsewhere.

Comment (Andreas-Koepke) [reply bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:22]: All of the other traits in that folder were also marked as internal so I just made it the same.

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+trait UserNameTrait
+{
+    /**
+     * @param int $user_id The User ID.
+     *
+     * @return array{username: string, fullname: string}
+     */
+    protected function getUserName(int $user_id): array
```
Comment (aryaantony92) [inline bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]:
```suggestion
    protected function getUserName(int $userId): array
```

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $user = User::getById($user_id);
```
Comment (aryaantony92) [inline bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]:
```suggestion
        $user = User::getById($userId);
```

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+        if (empty($user)) {
+            $data = [
+                'username' => '',
+                'fullname' => $this->trans('user_unknown'),
```
Comment (aryaantony92) [inline bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]:
```suggestion
                'userName' => '',
                'fullName' => $this->trans('user_unknown'),
```

Comment (aryaantony92) [reply bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]: Could you please change the 'username' and 'fullname' also to Camel case like the suggestion ? Thanks

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php (hunk 6 above)
Comment (brusch) [reply bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:22]: Agree, we should keep that one `@internal`.

File: bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php (hunk 9 above)
Comment (Andreas-Koepke) [reply bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php]: I have made this change and the associate changes to the controllers.

**Comment by aryaantony92** [comment]
@Andreas-Koepke I just tested the PR and it's working fine. But, what about moving the conditions like https://github.com/pimcore/pimcore/blob/a4fd44d71f805aa7e48f5475e76adcbcaff97d24/bundles/AdminBundle/public/js/pimcore/asset/asset.js#L495
https://github.com/pimcore/pimcore/blob/a4fd44d71f805aa7e48f5475e76adcbcaff97d24/bundles/AdminBundle/public/js/pimcore/asset/asset.js#L496 in Js files to the controller ? So, we need to pass only one variable from the controller (if full name exists, then full name otherwise the username). Since, we are using either full name/username in the js files, we don't need both of them in the js.

**Comment by Andreas-Koepke** [comment]
I just made a change to the UserNameTrait that checks if the user's full name is available. I have removed the check from JS files as it is no longer necessary.

At the moment, I have left both variables in the data returned by the controllers.

**Comment by aryaantony92** [comment]
@Andreas-Koepke Thanks a lot!

**Pull request merged**